lib/Dialect/RTG/Transforms/LowerValidateToLabelsPass.cpp

```diff
@@ -0,0 +1,64 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/RTG/IR/RTGOps.h"
+#include "circt/Dialect/RTG/Transforms/RTGPasses.h"
+#include "circt/Support/UnusedOpPruner.h"
+
+namespace circt {
+namespace rtg {
+#define GEN_PASS_DEF_LOWERVALIDATETOLABELSPASS
+#include "circt/Dialect/RTG/Transforms/RTGPasses.h.inc"
+} // namespace rtg
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+//===----------------------------------------------------------------------===//
+// Lower Validate To Labels Pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerValidateToLabelsPass
+    : public rtg::impl::LowerValidateToLabelsPassBase<
+          LowerValidateToLabelsPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+void LowerValidateToLabelsPass::runOnOperation() {
+  auto *rootOp = getOperation();
+  UnusedOpPruner pruner;
+
+  auto result = rootOp->walk([&](rtg::ValidateOp validateOp) -> WalkResult {
+    Location loc = validateOp.getLoc();
+    auto regOp = validateOp.getRef().getDefiningOp<rtg::FixedRegisterOp>();
+    if (!regOp)
+      return validateOp->emitError("could not determine register");
+
+    if (!validateOp.getId().has_value())
+      return validateOp.emitError("expected ID to be set");
+
+    OpBuilder builder(validateOp);
+    auto intrinsicLabel = validateOp.getRef().getType().getIntrinsicLabel(
+        regOp.getReg(), validateOp.getId().value());
+    Value lbl = builder.create<rtg::LabelDeclOp>(
+        loc, StringAttr::get(&getContext(), intrinsicLabel), ValueRange());
+    builder.create<rtg::LabelOp>(loc, rtg::LabelVisibility::global, lbl);
+    validateOp.getValue().replaceAllUsesWith(validateOp.getDefaultValue());
+
+    pruner.eraseNow(validateOp);
```

Comment:
nit: Strictly speaking it's not allowed to erase operations in walk other than the current operation (`validateOp`). Though I think it's always legal with the current implementation of walk specifically.